index/segment/types.go

```diff
@@ -136,3 +105,57 @@ type Filter struct {
 	// Note: RE2 only, no PCRE (i.e. no backreferences).
 	Regexp bool
 }
+
+// ImmutablePostingsList is a memory efficient collection of docIDs. It only permits
+// for read-only operations.
+type ImmutablePostingsList interface {
+	// Contains returns whether the specified id is contained in the set.
+	Contains(i DocID) bool
+
+	// IsEmpty returns whether the set is empty.
+	IsEmpty() bool
+
+	// Size returns the numbers of ids in the set.
+	Size() uint64
+
+	// Iter returns an iterator over the known values.
+	Iter() PostingsIter
+
+	// Clone returns a copy of the set.
+	Clone() PostingsList
+}
+
+// PostingsList is a memory efficient collection of docIDs.
+type PostingsList interface {
+	ImmutablePostingsList
+
+	// Insert inserts the given id to set.
+	Insert(i DocID)
+
+	// Intersect modifies the receiver set to contain only those ids by both sets.
```

Comment:
Would it be possible to move these methods out of the interface (i.e. `Intersect(pLists ImmutablePostingsList...) ImmutablePostingsList`)? Requiring it to mutate the underlying postings list might not make sense for all postings list implementations, and it will guarantee that we have to make a copy of at least one of the postings lists when responding to a query, but if we intersect them all together we can may be able to avoid any copy or allocate much less memory when we do.

Reply:
Lets defer this refactor till we have an additional PostingsList implementation/need to do this for other reasons (perf/allocation/...).